DOL.WHD.Section14c.Test/ActionFilters/LoggingFilterAttributeTests.cs

```diff
@@ -0,0 +1,75 @@
+﻿using DOL.WHD.Section14c.Log.ActionFilters;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.Web;
+using System.Net;
+using System.Web.Http.Controllers;
+using Moq;
+using System.Net.Http;
+using System.Text;
+using DOL.WHD.Section14c.Log.LogHelper;
+using System.Web.Http;
+using System;
+
+namespace DOL.WHD.Section14c.Log.ActionFilters.Tests
+{
+    [TestClass()]
+    public class LoggingFilterAttributeTests
+    {
+        private HttpActionContext httpActionContext;
+        private LoggingFilterAttribute filter;
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            var httpRequestMessage = new Mock<HttpRequestMessage>();
+            filter = new LoggingFilterAttribute();
+
+            HttpConfiguration config = new HttpConfiguration();
+            string controllerName = "UsersController";
+            Type controllerType = typeof(BaseApiController);
+
+            var httpActionDescriptor = new Mock<HttpActionDescriptor>();
+            httpActionDescriptor.Name = "Test";
+
+            httpActionContext = new HttpActionContext
+            {
+                ControllerContext = new HttpControllerContext
+                {
+                    Request = httpRequestMessage.Object as HttpRequestMessage,
+                    ControllerDescriptor = new HttpControllerDescriptor(config, controllerName, controllerType)
+                },
+                Response = new HttpResponseMessage
+                {
+                    StatusCode = HttpStatusCode.Conflict,
+                    Content = new StringContent("test")
+                },
+                ActionDescriptor = httpActionDescriptor.Object as HttpActionDescriptor
+            };
+        }
+
+
+        [TestMethod()]
+        public void LoggingFilterAttribute_AllowMultipleTest()
+        {
+            LoggingFilterAttribute actionFilter = new LoggingFilterAttribute();
+
+            Assert.IsTrue(actionFilter.AllowMultiple);
+        }
+
+        [TestMethod()]
+        public void LoggingFilterAttribute_OnActionExecutingTest()
+        {
+
+            filter.OnActionExecuting(httpActionContext);
```

Comment:
For this test, nothing is getting returned when it executed without error. There is no other parameter to pass. I can only confirm that it's successfully executed at this time :(. I will take any suggestions.

Reply:
I played around with it because I had an idea of how you'd test it:

1. In LoggingFilterAttribute.cs, move the `...Service.Replace()` line into a constructor
2. At the start of this test, create a mock of `ITraceWriter`
3. Call `...Service.Replace(typeof(ITraceWriter), mockedTrace.object)` to inject our mock into the `LoggingFilterAttribute` class
4. call `onActionExecuting`
5. verify that `mockedTrace.Info` was called with the expected arguments

Unfortunately, it turns out that Microsoft has "helpfully" provided extension methods for all of the functions on `ITraceWriter`, so moq can't do anything with them.  Since the only thing to be tested is verifying that `ITraceWriter.Info` is called, Microsoft has screwed us on this one.

I think this would require using Moles and I don't know enough about them to say more than that.  I'm also pretty unhappy with the way Microsoft has chosen to implement this, but there's nothing for any of us to do about that. 😆 

Thanks for taking another look, @dciceman.  For now, would you mind adding a couple of lines of comments just explaining that we can't test more than successful execution?  Maybe include a link to this comment (https://github.com/18F/dol-whd-14c/pull/316#discussion_r147143929) so anyone looking at it in the future can have context?

Reply:
@mgwalker @dciceman .Net Core seems to have better support for mocking + logging. Greg, I believe moles (along with Pex) is dead? Moq is what I have seen more of recently. No action needed ...